target/scripts/startup/setup.d/ldap.sh

```diff
@@ -2,80 +2,63 @@
-  DOVECOT_LDAP_MAPPING['DOVECOT_PASS_FILTER']="${DOVECOT_PASS_FILTER:="${DOVECOT_USER_FILTER}"}"
-
-  for VAR in "${!DOVECOT_LDAP_MAPPING[@]}"; do
-    export "${VAR}=${DOVECOT_LDAP_MAPPING[${VAR}]}"
-  done
-
-  _replace_by_env_in_file 'DOVECOT_' '/etc/dovecot/dovecot-ldap.conf.ext'
+  local DOVECOT_PASS_FILTER="${DOVECOT_PASS_FILTER:="${DOVECOT_USER_FILTER}"}"
+  _create_config_dovecot
 
   _log 'trace' 'Enabling Dovecot LDAP authentication'
 
   sed -i -e '/\!include auth-ldap\.conf\.ext/s/^#//' /etc/dovecot/conf.d/10-auth.conf
   sed -i -e '/\!include auth-passwdfile\.inc/s/^/#/' /etc/dovecot/conf.d/10-auth.conf
 
-  _log 'trace' "Configuring LDAP"
+  # shellcheck disable=SC2016
+  sed -i 's|mydestination = \$myhostname, |mydestination = |' /etc/postfix/main.cf
 
-  if [[ -f /etc/postfix/ldap-users.cf ]]; then
-    postconf 'virtual_mailbox_maps = ldap:/etc/postfix/ldap-users.cf'
-  else
-    _log 'warn' "'/etc/postfix/ldap-users.cf' not found"
-  fi
+  return 0
+}
 
-  if [[ -f /etc/postfix/ldap-domains.cf ]]; then
-    postconf 'virtual_mailbox_domains = /etc/postfix/vhost, ldap:/etc/postfix/ldap-domains.cf'
-  else
-    _log 'warn' "'/etc/postfix/ldap-domains.cf' not found"
-  fi
+# Generates a config from an ENV template while layering several other sources
+# into a single temporary file, used as input into `_cleanse_config` which
+# prepares the final output config.
+function _create_config_dovecot() {
+  _cleanse_config '=' <(cat 2>/dev/null \
+    <(_template_with_env 'LDAP_' /etc/dms/ldap/dovecot.base) \
+    /tmp/docker-mailserver/ldap/dovecot.conf \
+    <(_template_with_env 'DOVECOT_' /etc/dms/ldap/dovecot.tmpl) \
+  ) > /etc/dovecot/dovecot-ldap.conf.ext
+}
 
-  if [[ -f /etc/postfix/ldap-aliases.cf ]] && [[ -f /etc/postfix/ldap-groups.cf ]]; then
-    postconf 'virtual_alias_maps = ldap:/etc/postfix/ldap-aliases.cf, ldap:/etc/postfix/ldap-groups.cf'
-  else
-    _log 'warn' "'/etc/postfix/ldap-aliases.cf' and / or '/etc/postfix/ldap-groups.cf' not found"
-  fi
+function _create_config_postfix() {
+local QUERY_KIND=${1:?QUERY_KIND is required in _create_config_postfix}
+  local LDAP_CONFIG_FILE="/etc/postfix/ldap/${QUERY_KIND}.cf"
 
-  # shellcheck disable=SC2016
-  sed -i 's|mydestination = \$myhostname, |mydestination = |' /etc/postfix/main.cf
+  _cleanse_config '=' <(cat 2>/dev/null \
+    <(_template_with_env 'LDAP_' /etc/dms/ldap/postfix.base) \
+    "/tmp/docker-mailserver/ldap-${QUERY_KIND}.cf" \
+    <(_template_with_env 'POSTFIX_' /etc/dms/ldap/postfix.tmpl) \
+    <(_template_with_env "POSTFIX_${QUERY_KIND^^}_" /etc/dms/ldap/postfix.tmpl) \
+  ) > "${LDAP_CONFIG_FILE}"
 
-  return 0
+  # Opt-out of generated config if `query_filter` was not configured:
+  if ! grep -q '^query_filter =' "${LDAP_CONFIG_FILE}"; then
+    _log 'warn' "'${LDAP_CONFIG_FILE}' is missing the 'query_filter' setting - disabling"
+
+    sed -i "s/$(_escape_for_sed "${LDAP_CONFIG_FILE}")//" /etc/postfix/main.cf
+  fi
```

Comment:
This won't handle `smtpd_sender_login_maps` in `spoofing.sh`.

Technically we could add our own variable to `main.cf` for this like we've done with milters and restrictions, so that we could disable that.

Not sure if removing the file from the `main.cf` setting like this is the correct call either, but the alternative would be restoring the conditionals for retaining the original file based lookups that may not exist with `ACCOUNT_PROVISIONER=LDAP`, so perhaps this opt-out approach is the way to go?

Reply:
The LDAP support isn't as complicated as originally thought:
- LDAP service has the data and receives a search query with 1 or more keys (attributes) with some value or pattern to match.
- The query is like a conditional statement in that sense and supports `&` (aka `&&`) or `|` (aka (`||`). That's good enough for DMS at least.
- Postfix has the `ldap:/` lookup table type. Same config for each separate file that we assign in `main.cf`, they mostly only differ by the `query_filter` (search query).

Technical details aside, they function and work the same as our text file counterparts, the query just makes the content a bit more dynamic instead of the left-side column lookup in `text:/` tables.

> lack a bit of context (due to a lack of time)

No worries, I understand 👍 

I'll be running out of time soon enough myself, just trying to avoid the LDAP support rotting away :)